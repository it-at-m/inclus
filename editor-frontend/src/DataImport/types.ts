import { Place } from "../types";

export enum ImportErrorLevel {
  ERROR = "ERROR",
  WARNING = "WARNING",
}

export type ValidationResult = {
  isValid: boolean;
  message?: string;
};

export type Validator = (value: any) => ValidationResult;

export type ValidatorByIndex = (index: number) => Validator;

export type ValidatorByFieldName = (fieldName: string) => Validator;

export type FieldConfig = {
  fieldName: keyof Place;
  getter: any;
  validator?: ValidatorByIndex | ValidatorByFieldName | Validator;
  errorLevel: ImportErrorLevel;
};
