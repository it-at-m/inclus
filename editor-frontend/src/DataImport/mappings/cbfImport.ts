import {
  getAdditionalField,
  getAdditionalFieldNumberOrEmpty,
  getAddress,
  getEuroKey,
  getNumberByArrayIndex,
  getPhotoURL,
  getPlanURL,
  getPosition,
  getStringByArrayIndex,
} from "../getters";
import { FieldConfig, ImportErrorLevel } from "../types";
import {
  addressValidator,
  euroKeyValidator,
  gripValidator,
  imageURLValidator,
  nonEmptyStringValidator,
  positionValidator,
  numberByIndexValidator,
  numberAdditionalFieldValidator,
  nonEmptyAdditionalFieldValidator,
} from "../validators";

/*
 This mapping descriptor is for data that we receive from CBF (https://www.cbf-muenchen.de/) and it assumes the following
 order of the csv headers:
 ['id', 'title', 'alias', 'type', 'catid', 'short_description', 'description', 'generate_coordinates', 'geodata', 'latitude', 'longitude',
 'geo_x', 'geo_y', 'geo_z', 'zoom', 'street', 'street_no', 'addr_addinfo', 'zip_code', 'city', 'region', 'country', 'phone', 'url', 'email',
 'additional_fields', 'images', 'icons', 'state', 'created', 'created_by', 'modified', 'modified_by', 'publish_up', 'publish_down',
 'checked_out', 'checked_out_time', 'ordering', 'metakey', 'metadesc', 'metadata', 'access', 'hits', 'featured', 'language', 'version', 'params']
 */

export const mappings: Array<FieldConfig> = [
  {
    fieldName: "title",
    getter: getStringByArrayIndex(1),
    validator: nonEmptyStringValidator(1),
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "short_description",
    getter: getStringByArrayIndex(5),
    validator: nonEmptyStringValidator(5),
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "photo",
    getter: getPhotoURL,
    validator: imageURLValidator,
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "plan",
    getter: getPlanURL,
    validator: imageURLValidator,
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "position",
    getter: getPosition,
    validator: positionValidator,
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "address",
    getter: getAddress,
    validator: addressValidator,
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "eurokey",
    getter: getEuroKey,
    validator: euroKeyValidator,
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "ramp_steepness",
    getter: getAdditionalFieldNumberOrEmpty("wc_ramp_incline"),
    validator: numberAdditionalFieldValidator("wc_ramp_incline"),
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "ramp_length",
    getter: getAdditionalFieldNumberOrEmpty("wc_ramp_length"),
    validator: numberAdditionalFieldValidator("wc_ramp_length"),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "city",
    getter: getStringByArrayIndex(19),
    validator: nonEmptyStringValidator(19),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "modified",
    getter: getStringByArrayIndex(31),
    validator: nonEmptyStringValidator(31),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "direction",
    getter: getStringByArrayIndex(6),
    validator: nonEmptyStringValidator(6),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "zip_code",
    getter: getNumberByArrayIndex(18),
    validator: numberByIndexValidator(18),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "id",
    getter: getNumberByArrayIndex(0),
    validator: numberByIndexValidator(0),
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "cbf_id",
    getter: getNumberByArrayIndex(0),
    validator: numberByIndexValidator(0),
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "wc_details",
    getter: getAdditionalField("wc_addinfo"),
    validator: nonEmptyAdditionalFieldValidator("wc_addinfo"),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "wc_accessright",
    getter: getAdditionalField("wc_accessright"),
    // validator: numberAdditionalFieldValidator("wc_accessright"),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "wc_accessleft",
    getter: getAdditionalField("wc_accessleft"),
    // validator: numberAdditionalFieldValidator("wc_accessleft"),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "interior_description",
    getter: getAdditionalField("wc_roominfos"),
    validator: nonEmptyAdditionalFieldValidator("wc_roominfos"),
    errorLevel: ImportErrorLevel.WARNING,
  },
  {
    fieldName: "grip",
    getter: getAdditionalField("wc_wheregrip"),
    validator: gripValidator,
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "door_width",
    getter: getAdditionalField("wc_door_width"),
    validator: numberAdditionalFieldValidator("wc_door_width"),
    errorLevel: ImportErrorLevel.ERROR,
  },
  {
    fieldName: "access",
    getter: getAdditionalField("wc_infos"),
    validator: nonEmptyAdditionalFieldValidator("wc_infos"),
    errorLevel: ImportErrorLevel.WARNING,
  },
];
