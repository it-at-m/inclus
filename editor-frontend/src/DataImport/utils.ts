import {
  FieldConfig,
  ImportErrorLevel,
  ValidationResult,
  Validator,
} from "./types";
import { Place } from "../types";

export function strip(html: string) {
  // note: this function only works when executed in a browser context
  let doc = new DOMParser().parseFromString(html, "text/html");
  return doc.body.textContent || "";
}

export const createPlaceFromCSVLine = async (
  csvLine: Array<string>,
  importConfig: Array<FieldConfig>
) => {
  let place: Partial<Place> = {};
  const errors: Array<string> = [];
  const warnings: Array<string> = [];
  for (let fieldConfig of importConfig) {
    let value = await fieldConfig.getter(csvLine);

    if (fieldConfig.validator) {
      let validationResult: ValidationResult = (
        fieldConfig.validator as Validator
      )(value);

      const errorLevel = fieldConfig.errorLevel;

      if (!validationResult.isValid) {
        if (errorLevel === ImportErrorLevel.ERROR && validationResult.message) {
          errors.push(validationResult.message);
        }
        if (
          errorLevel === ImportErrorLevel.WARNING &&
          validationResult.message
        ) {
          warnings.push(validationResult.message);
        }
      }
    }
    place[fieldConfig.fieldName] = value;
  }
  return { place, errors, warnings };
};

// TODO: remove this to core package utility

const asin = Math.asin;
const cos = Math.cos;
const sin = Math.sin;
const sqrt = Math.sqrt;
const PI = Math.PI;
const R = 6378137;

function squared(x: number) {
  return x * x;
}
function toRad(x: number) {
  return (x * PI) / 180.0;
}
function hav(x: number) {
  return squared(sin(x / 2));
}

function haversineDistance(a: [number, number], b: [number, number]) {
  const aLat = toRad(a[1]);
  const bLat = toRad(b[1]);
  const aLng = toRad(a[0]);
  const bLng = toRad(b[0]);

  const ht = hav(bLat - aLat) + cos(aLat) * cos(bLat) * hav(bLng - aLng);
  return 2 * R * asin(sqrt(ht));
}

export function computeClosestPoints(
  dataPoints: Array<Place>,
  targetPoint: Place,
  numClosest = 5
) {
  const distances = dataPoints.map((point) => ({
    point,
    distance:
      point !== targetPoint
        ? haversineDistance(targetPoint.position, point.position)
        : Infinity,
  }));

  distances.sort((a, b) => a.distance - b.distance);

  return distances.slice(0, numClosest);
}
