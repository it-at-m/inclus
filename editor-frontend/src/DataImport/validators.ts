import { Validator, ValidatorByFieldName, ValidatorByIndex } from "./types";
import { Grip } from "../types";

export const nonEmptyStringValidator: ValidatorByIndex =
  (index: number) => (value: string) => {
    if (!value || value.length === 0) {
      return {
        isValid: false,
        message: "Value at index " + index + " cannot be empty",
      };
    } else {
      return { isValid: true };
    }
  };

export const nonEmptyAdditionalFieldValidator: ValidatorByFieldName =
  (fieldName: string) => (value: string) => {
    if (!value || value.length === 0) {
      return {
        isValid: false,
        message: "Value at additional field " + fieldName + " cannot be empty",
      };
    } else {
      return { isValid: true };
    }
  };

export const numberByIndexValidator: ValidatorByIndex =
  (index: number) => (value: number) => {
    if (!value || Number.isNaN(value)) {
      return {
        isValid: false,
        message:
          "Value at index " +
            index +
            " must not be empty and must be a number, provided " +
            value || "empty value",
      };
    } else {
      return { isValid: true };
    }
  };

export const numberAdditionalFieldValidator: ValidatorByFieldName =
  (fieldName) => (value: string) => {
    if (!value) {
      return {
        isValid: false,
        message:
          "Value at additional fields " + fieldName + " must not be empty",
      };
    }

    if (Number.isNaN(parseFloat(value))) {
      return {
        isValid: false,
        message:
          "Value at additional fields " +
          fieldName +
          " must be a number, provided " +
          value,
      };
    }

    return { isValid: true };
  };

export const positionValidator = (value: Array<any>) => {
  let message = "";
  let isValid = true;
  if (Number.isNaN(value[0])) {
    isValid = false;
    message +=
      "Value at index 9 must be a number to represent correct latitude. ";
  }
  if (Number.isNaN(value[1])) {
    isValid = false;
    message +=
      "Value at index 10 must be a number to represent correct longitude. ";
  }

  if (value[0] === 0 || value[1] === 0) {
    isValid = false;
    message = "Position values (indexes 9 and 10) must not be empty or 0";
  }
  return { isValid, message };
};

export const addressValidator = (value: string) => {
  if (!value || value.length === 0 || value === " ") {
    return {
      isValid: false,
      message:
        "Values at indexes 15 and 16 must not be empty to form a valid address",
    };
  } else {
    return { isValid: true };
  }
};

export const gripValidator = (value: string) => {
  let gripValues = "";
  for (let validGrip in Grip) {
    if (value === validGrip) {
      return { isValid: true };
    }
    gripValues += validGrip + ", ";
  }
  return {
    isValid: false,
    message:
      "Value of additional field wc_wheregrip must be one of: " +
      gripValues +
      " provided " +
      value,
  };
};

export const euroKeyValidator: Validator = (value: string) => {
  if (value && ["0", "1", "ka"].includes(value)) {
    return { isValid: true };
  }
  return {
    isValid: false,
    message:
      "Value of additional field wc_euro must be either 0, 1 or ka provided " +
        value || "empty value",
  };
};

export const imageURLValidator = (value: string) => {
  if (value && value.length > 0 && value.match(/[^/]+(jpg|png)$/)) {
    return { isValid: true };
  } else {
    return {
      isValid: false,
      message:
        "Values of image filenames must not be empty and must be valid image filenames",
    };
  }
};
