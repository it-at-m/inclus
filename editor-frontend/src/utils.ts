import { Place } from "./types";

export const filterPlaces = (filterString: string, places: Place[]) => {
  return places.filter(
    (place: Place) =>
      place.title.toLowerCase().indexOf(filterString.toLowerCase()) !== -1 ||
      place.short_description
        .toLowerCase()
        .indexOf(filterString.toLowerCase()) !== -1 ||
      place.id + "" === filterString.toLowerCase()
  );
};
