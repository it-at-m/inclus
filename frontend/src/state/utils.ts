import { Place, PlaceAndDistance } from "../types/place";
import { DoorWidths, EuroKeyValues, RampValues } from "../types/enums";
import { FilterValue, SortableListItem } from "./types";
import { LatLng, LatLngExpression } from "leaflet";

export const filterPlaces = (places: Place[], filterValue: FilterValue) => {
	let filteredPlaces = places;
	if (
		filterValue.euroKey !== undefined &&
		filterValue.euroKey !== EuroKeyValues.NOT_SELECTED
	) {
		filteredPlaces = filteredPlaces.filter(
			(place) => Number(place.eurokey) === filterValue.euroKey
		);
	}
	if (
		filterValue.doorWidth !== undefined &&
		filterValue.doorWidth !== DoorWidths.NOT_SELECTED
	) {
		filteredPlaces = filteredPlaces.filter(
			(place) => Number(place.door_width) >= filterValue.doorWidth
		);
	}
	if (
		filterValue.rampVal !== undefined &&
		filterValue.rampVal !== RampValues.NOT_SELECTED
	) {
		filteredPlaces = filteredPlaces.filter((place) => {
			return Number(place.ramp_steepness) <= filterValue.rampVal;
		});
	}
	return filteredPlaces;
};

export const filterByDistance = (
	placeList: Place[],
	locationMeLatLng: LatLng | null,
	targetDistance: number // m
): PlaceAndDistance[] => {
	if (!locationMeLatLng) {
		return [];
	}

	const filteredList: PlaceAndDistance[] = [];

	placeList.forEach((item) => {
		const locationLatLng = new LatLng(item.position[0], item.position[1]);

		// Use the distanceTo method from Leaflet to calculate the distance between two locations.
		const distance = Math.round(
			locationMeLatLng.distanceTo(locationLatLng)
		);
		if (distance <= targetDistance) {
			filteredList.push({ place: item, distance });
		}
	});
	return filteredList;
};

export const sortAndSliceByDistance = <T extends SortableListItem>(
	list: T[],
	limit: number,
	mapCenter: LatLngExpression
): T[] => {
	if (!list || list.length === 0) {
		return [];
	}

	if (!mapCenter) return [];

	const sortedList = list.sort((a, b) => {
		const aLatLng = new LatLng(
			a.position?.[0] || a.y!,
			a.position?.[1] || a.x!
		);
		const bLatLng = new LatLng(
			b.position?.[0] || b.y!,
			b.position?.[1] || b.x!
		);

		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		// @ts-ignore
		return aLatLng.distanceTo(mapCenter) - bLatLng.distanceTo(mapCenter);
	});

	return sortedList.slice(0, limit);
};
