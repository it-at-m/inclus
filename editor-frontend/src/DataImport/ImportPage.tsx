import React, {
  ChangeEvent,
  ChangeEventHandler,
  RefObject,
  useRef,
  useState,
} from "react";
import Papa from "papaparse";
import { saveAs } from "file-saver";
import { computeClosestPoints, createPlaceFromCSVLine } from "./utils";
import { mappings } from "./mappings/cbfImport";
import { Place } from "../types";

const numberOfClosestToilets = 3;

const ImportPage = () => {
  const [numberOfPlaces, setNumberOfPlaces] = useState(0);
  const hiddenFileInput: RefObject<HTMLInputElement> =
    useRef<HTMLInputElement>(null);

  const [loading, setLoading] = useState(false);

  const [notifications, setNotifications] = useState<Array<any>>([]);

  const handleFileInput: ChangeEventHandler<HTMLInputElement> = async (
    event: ChangeEvent<HTMLInputElement>
  ) => {
    if (event.target.files) {
      const file = event.target.files[0];
      setLoading(true);

      Papa.parse(file, {
        complete: async function (results) {
          setLoading(false);

          const newPlaces: Array<Place> = [];

          const allNotifications: Array<any> = [];

          for (let i = 1; i < results.data.length; i++) {
            // first csv row contains headers
            const rawPlace: Array<any> = results.data[i] as Array<any>;

            let parserNotifications: any = {};
            if (rawPlace.length > 1) {
              const parsingResult = await createPlaceFromCSVLine(
                rawPlace,
                // for now we only have one type of import from cbf, if we get various csv formats,
                // the mappings object should be chosen based on the header list
                // (compare the headers received and choose the right mappings from available)
                mappings
              );

              if (
                parsingResult.errors.length > 0 ||
                parsingResult.warnings.length > 0
              ) {
                parserNotifications.place = parsingResult.place.title;
                parserNotifications.placeId = parsingResult.place.id;
                parserNotifications.errors = parsingResult.errors;
                parserNotifications.warnings = parsingResult.warnings;
                allNotifications.push(parserNotifications);
              }

              if (parsingResult.errors.length === 0)
                newPlaces.push(parsingResult.place as Place); // it is safe type conversion here, because we know that if there were no errors Place object was constructed complete from csvLine, but I'd like a more  robust check here
            }
          }

          for (const point of newPlaces) {
            const closestPoints = computeClosestPoints(
              newPlaces,
              point,
              numberOfClosestToilets
            );

            point["closest_toilets"] = closestPoints.map((cp) => ({
              id: cp.point.id,
              distance: Math.round(cp.distance),
            }));
          }

          let updatedJSON = JSON.stringify(newPlaces, null, 2);

          // TODO: remove this completely, only pass strings through json as should be
          const regex = /"(-|)([0-9]+(?:\.[0-9]+)?)"/g;
          updatedJSON = updatedJSON.replace(regex, "$1$2");

          const fileToSave = new Blob([updatedJSON], {
            type: "application/json",
          });

          setNumberOfPlaces(newPlaces.length);
          setNotifications(allNotifications);
          saveAs(fileToSave, "cache-munich.json");
        },
      });
    }
  };

  const handleChooseFile = () => {
    if (hiddenFileInput && hiddenFileInput.current)
      hiddenFileInput.current.click();
  };

  return (
    <div>
      <h3>Import new data</h3>
      <button className="fileUploadButton" onClick={handleChooseFile}>
        {loading ? "..." : "Choose CSV file"}
      </button>
      <input
        type="file"
        ref={hiddenFileInput}
        accept=".csv"
        onChange={handleFileInput}
      />
      {numberOfPlaces !== 0 && (
        <span>
          {numberOfPlaces} toilets imported, {notifications.length} toilets
          contain errors or warnings.
        </span>
      )}
      {notifications.map((notification) => {
        return (
          <div key={notification.placeId}>
            <h4>{notification.place}</h4>
            <h5>
              {notification.errors.length > 0 &&
                "Object not imported due to errors"}
            </h5>
            {notification.errors.map((error: string, index: number) => (
              <p key={notification.placeId + "_errors_" + index}>{error}</p>
            ))}
            <h5>
              {notification.warnings.length > 0 &&
                (notification.errors.length > 0
                  ? "Warnings"
                  : "Object imported with warnings")}
            </h5>
            {notification.warnings.map((warning: string, index: number) => (
              <p key={notification.placeId + "_warnings_" + index}>{warning}</p>
            ))}
          </div>
        );
      })}
    </div>
  );
};

export default ImportPage;
