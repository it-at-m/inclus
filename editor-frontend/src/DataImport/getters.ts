import { strip } from "./utils";

export const getStringByArrayIndex =
  (arrayIndex: number) => (csvLine: Array<string>) => {
    return strip(csvLine[arrayIndex]);
  };

export const getNumberByArrayIndex =
  (arrayIndex: number) => (csvLine: Array<string>) => {
    return Number(csvLine[arrayIndex]);
  };

export const getAdditionalField =
  (additionalFieldName: string) => async (csvLine: Array<string>) => {
    const additionalFields = await JSON.parse(csvLine[25]);
    return strip(additionalFields[additionalFieldName]);
  };

export const getEuroKey = async (csvLine: Array<string>) => {
  const additionalFields = await JSON.parse(csvLine[25]);
  let value = strip(additionalFields["wc_euro"]);
  if (value === "undefined") value = "ka";
  return value;
};

export const getAdditionalFieldNumberOrEmpty =
  (additionalFieldName: string) => async (csvLine: Array<string>) => {
    const additionalFields = await JSON.parse(csvLine[25]);
    let value = strip(additionalFields[additionalFieldName]);
    if (value === undefined || !value) value = "0";
    return value;
  };

export const getPhotoURL = async (csvLine: Array<string>) => {
  const images = await JSON.parse(csvLine[26]);
  return (images.image_intro as string).replace("images/", "watermarked/");
};

export const getPlanURL = async (csvLine: Array<string>) => {
  const additionalFields = await JSON.parse(csvLine[25]);
  return "watermarked/wc-plan/" + additionalFields.wc_plan + ".jpg";
};

export const getPosition = (csvLine: Array<string>) => {
  return [Number(csvLine[9]), Number(csvLine[10])];
};

export const getAddress = (csvLine: Array<string>) => {
  return csvLine[15] + " " + csvLine[16];
};
