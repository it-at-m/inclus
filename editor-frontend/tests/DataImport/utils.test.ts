import { expect, test } from "vitest";
import { createPlaceFromCSVLine } from "../../src/DataImport/utils";
import {
  invalidAdditionalFieldsCSVLineArray,
  invalidCSVLineArray,
  validCSVLineArray,
  warningsCSVLineArray,
} from "../mocks/DataImport/cbfMocks";
import { mappings as CBFMappings } from "../../src/DataImport/mappings/cbfImport";

test("parses a valid cbf export csv line into a valid place object", async () => {
  const { place, errors, warnings } = await createPlaceFromCSVLine(
    validCSVLineArray,
    CBFMappings
  );

  expect(errors).toHaveLength(0);
  expect(warnings).toHaveLength(0);
  expect(place).toMatchSnapshot();
});

test("reports errors if no id, title, position and short_description was submitted", async () => {
  const { errors } = await createPlaceFromCSVLine(
    invalidCSVLineArray,
    CBFMappings
  );
  expect(errors).toMatchSnapshot();
});

test("reports errors for eurokey, ramp, grip and door_width values", async () => {
  const { errors } = await createPlaceFromCSVLine(
    invalidAdditionalFieldsCSVLineArray,
    CBFMappings
  );
  expect(errors).toMatchSnapshot();
});

test("parses a cbf csv line with warnings", async () => {
  const { warnings } = await createPlaceFromCSVLine(
    warningsCSVLineArray,
    CBFMappings
  );
  expect(warnings).toMatchSnapshot();
});
