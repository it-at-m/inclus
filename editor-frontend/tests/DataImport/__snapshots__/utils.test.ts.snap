// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parses a cbf csv line with warnings 1`] = `
[
  "Values of image filenames must not be empty and must be valid image filenames",
  "Values at indexes 15 and 16 must not be empty to form a valid address",
  "Value at index 19 cannot be empty",
  "Value at index 31 cannot be empty",
  "Value at index 6 cannot be empty",
  "Value at index 18 must not be empty and must be a number, provided 0",
  "Value at additional field wc_addinfo cannot be empty",
  "Value at additional field wc_roominfos cannot be empty",
  "Value at additional field wc_infos cannot be empty",
]
`;

exports[`parses a valid cbf export csv line into a valid place object 1`] = `
{
  "access": "Das WC-Becken ist schräg links gegenüber von der Tür, links neben dem Waschbecken.",
  "address": "Somestr. 2",
  "cbf_id": 5,
  "city": "Unterhaching",
  "direction": "Das rollstuhlgerechte WC in einem Freibad",
  "door_width": "75",
  "eurokey": "1",
  "grip": "lr",
  "id": 5,
  "interior_description": "Zwischen Umkleide und Duschbad ist eine Schiebetür, die nach links öffnet.",
  "modified": "2023-08-29 11:15:20",
  "photo": "watermarked/wc-af/ba-unterh1WC1f1.jpg",
  "plan": "watermarked/wc-plan/ba-unterh1WC3c.jpg",
  "position": [
    48.0619761,
    11.6208199,
  ],
  "ramp_length": "0",
  "ramp_steepness": "0",
  "short_description": "Bedingt öffentliche Toilette mit Dusche in einem Freibad, mit Euroschloss.",
  "title": "WC Freibad",
  "wc_accessleft": "75",
  "wc_accessright": "105",
  "wc_details": "",
  "zip_code": 82008,
}
`;

exports[`reports errors for eurokey, ramp, grip and door_width values 1`] = `
[
  "Value of additional field wc_euro must be either 0, 1 or ka provided vroum",
  "Value at additional fields wc_ramp_incline must be a number, provided some percentage",
  "Value of additional field wc_wheregrip must be one of: lr, l, rfl, flr, fr, none, lfr, r, fl,  provided lrrrrr",
  "Value at additional fields wc_door_width must be a number, provided some centimeters",
]
`;

exports[`reports errors if no id, title, position and short_description was submitted 1`] = `
[
  "Value at index 1 cannot be empty",
  "Value at index 5 cannot be empty",
  "Value at index 9 must be a number to represent correct latitude. Value at index 10 must be a number to represent correct longitude. ",
  "Value at index 0 must not be empty and must be a number, provided 0",
  "Value at index 0 must not be empty and must be a number, provided 0",
]
`;
